.commitlintrc

```diff
@@ -0,0 +1,3 @@
+{
+  extends: ["@commitlint/config-conventional"]
```

Comment:
I believe you wanted to customise the config more. :wink:

Consider this comment just a note without any action required.

Reply:
i actually like the default conventional config, but i'm happy to config more. what configurations would you like to see differently configured?

Reply:
Well, we might use the default config, I just thought you wanted to configure it more.

I prefer to list all allowed types and scopes, and use the [`selective-scope`](www.npmjs.com/package/commitlint-plugin-selective-scope) plugin to make sure that some scopes are used only for specific types (e.g. all package names could be used as scopes for any type but `chore`; for `chore` type, I only use `[null, 'build', 'ci', 'deps']` where `null` means no scope).

I could show you my `commitlint.config.js` if you want.

Reply:
i'm happy to take a look, i don't think i'd agree that we should use selective-scope without good reason

Reply:
<details>
  <summary><code>commitlint.config.js</code></summary>

```js
const {existsSync, readdirSync} = require('node:fs')

const appsLibsScopes = [null]

// For `chore`, these are the only allowed scopes.
const choreScopes = [null, 'build', 'ci', 'deps']

// List all apps and libraries, and all Nest modules in them
// Note: I use Nx monorepo and Nest.js app.
// Note: I still want to fix this to select Nest modules only. For now, all it does is select subfolders under `{apps,libs}/$app/src`. I also want to get the apps and libs list programmatically using Nx.
// Examples: `app1`, `app2`, `app1|module1`.
;[`${__dirname}/apps`, `${__dirname}/libs`].forEach(f => {
  if (existsSync(f)) {
    readdirSync(f, {withFileTypes: true})
      .filter(dirent => dirent.isDirectory())
      .map(dirent => dirent.name)
      .forEach(s => {
        if (existsSync(`${f}/${s}/src`)) {
          const modules = readdirSync(`${f}/${s}/src`, {withFileTypes: true})
            .filter(dirent => dirent.isDirectory())
            .map(dirent => dirent.name)

          appsLibsScopes.push(s, ...modules.map(m => `${s}|${m}`))
        } else {
          appsLibsScopes.push(s)
        }
      })
  }
})

// `docs` location
// Note: That I am in the process of moving the root-level `docs` folder to a Vitepress app under `apps/`.
const docsFolder = `${__dirname}/docs`

// List app docs subfolders which are treated the same way as the modules of apps/libs.
// Examples: `docs`, `docs/contribution`, `docs/usage`.
const docsScopes = [
  null,
  ...(existsSync(docsFolder)
    ? readdirSync(docsFolder, {withFileTypes: true}).map(i =>
        i.isFile() ? i.name.replace(/\.md/, '') : i.name
      )
    : [])
]

// Define all allowed scopes for each type
// Note: I know I could simplify this using a loop/`Array#map()`, however, this is more readable.
const typesEnumScoped = {
  // Dev tools related stuff, not related to the code itself
  // Commits, that affect build components like build tool, CI pipeline, dependencies, project version, ...
  // Miscellaneous commits, e.g. modifying `.gitignore`
  chore: choreScopes,

  // Commits, that affect documentation only
  docs: docsScopes,

  // Commits, that improves an existing feature (minor version bump)
  enh: appsLibsScopes,

  // Commits, that adds a new feature (minor version bump)
  feat: appsLibsScopes,

  // Commits that fix a bug (patch version bump)
  fix: appsLibsScopes,

  // Commits are special `refactor` commits, that improve performance
  perf: appsLibsScopes,

  // Commits, that rewrite/restructure your code, however, does not change any behaviour
  refactor: appsLibsScopes,

  // Commits that revert previous commits
  revert: appsLibsScopes,

  // Commits that do not affect the meaning (white-space, formatting, missing semicolons, etc)
  style: appsLibsScopes,

  // Commits, that add missing tests or correcting existing tests
  test: appsLibsScopes
}

module.exports = {
  extends: ['@commitlint/config-conventional'],
  formatter: '@commitlint/format',
  // TODO: Change this URL to our Wiki page with the description of the commit message format.
  helpUrl:
    'https://github.com/conventional-changelog/commitlint/#what-is-commitlint',
  parserPreset: {
    parserOpts: {
      issuePrefixes: ['#']
    }
  },
  plugins: ['selective-scope'],
  rules: {
    'body-case': [2, 'always', 'sentence-case'],
    'body-empty': [0, 'never'],
    'body-full-stop': [2, 'always', '.'],
    'body-leading-blank': [2, 'always'],
    'body-max-line-length': [0, 'never', 0],
    'footer-empty': [0, 'never'],
    'footer-leading-blank': [2, 'always'],
    'footer-max-line-length': [0, 'never', 0],
    'header-full-stop': [2, 'never', '.'],
    'header-max-length': [0, 'never', 0],
    'scope-enum': [
      2,
      'always',
      // From `commitlint` point of view, allow all scopes regardless of the type
      [...appsLibsScopes, ...choreScopes, ...docsScopes].filter(i => i !== null)
    ],
    // Use `selective-scope` to make sure that the scopes are only allowed when they are defined for a particular type
    'selective-scope': [2, 'always', typesEnumScoped],
    'subject-case': [1, 'always', 'lower-case'],
    'subject-empty': [2, 'never'],
    'subject-full-stop': [2, 'never', '.'],
    'type-case': [2, 'always', 'lower-case'],
    'type-enum': [2, 'always', Object.keys(typesEnumScoped)]
  }
}
```

</details>

Reply:
> i don't think i'd agree that we should use selective-scope without good reason

And I don’t want to force you to use it. All I wanted to say is that without it, `commitlint` allows any `type` and `scope` matches which IMHO might not be what one wants.

Reply:
that said, you do give a really good template even if we just want to do those 3 for now, and otherwise allowing scopes as empty for other files... what do you think of that?

Reply:
> unless we just want specific scopes for constants, models, and calendars, but i don't feel those would be significantly impactful right now

If you look at my commit messages in romcal, for scope of `feat` and `fix` types I try to use the following:

- `l10n`: related to l10n changes either of multiple locales or unrelated to a single l10n file;
- `calendar`: related to calendar changes either of multiple calendars or unrelated to a single l10n files;
- `$localeId` (e.g. `la` or `en`): related to a single locale;
- `$calendarId` (e.g. `general` or `europe` or `slovakia`): related to a single calendar;

For style changes, I used to use `chore(style)`, however, not anymore. Now, I use `style` type with either no scope or one of the scopes listed above.

The rest could be the same as I suggested in the `commitlint.config.js` … unless you disagree.

Reply:
> I'm fine with having restricted scopes, but not restricted types.

I’m not sure I understand you, however, both types and scopes are restricted by `commitlint` by default. :thinking:

Reply:
i think i'm willing to adapt some of what you have into a config for us. I'll work on that in a few hours. I misread 

> for scope of `feat` and `fix` types I try to use

to mean something you are not saying. Thank you for your clarification.

Reply:
> i think i'm willing to adapt some of what you have into a config for us. I'll work on that in a few hours.

Okay, I’ll review it then, probably tomorrow CET/UTC+01.

> I misread […]

Well, yeah, I wrote it fast. All I wanted to say by _for scope of `feat` and `fix` types I try to use_ is that for those types I have already used those scopes, however, I didn’t say anything about other types which should be kept as is, or possibly, we might want to move feature enhancements to `enh` from `feat` and keep `feat` for new features only. We also might want to merge/reduce `build`, `chore` and `ci` (I prefer using only `chore` with `build` and `ci` to be scopes for `chore`, however, we could keep them as they are).

Reply:
i think, `feat` should stay to cover both scenarios as-is, as they both produce the same minor version bump. Would you see `enh` as a patch rather than a minor version change?

for `build`/`chore`/`ci`, i'm down reducing chore out and only allowing `build` and `ci` for specific differences, as i think anything that's a `chore` in this case would really be either a `build` or `ci` change, and `chore` could be eliminated entirely. I'm open to your opinion on this.

Reply:
> i think, `feat` should stay to cover both scenarios as-is, as they both produce the same minor version bump. Would you see `enh` as a patch rather than a minor version change?

From release point of view, both `feat` and `enh` would cause a `minor` bump, however, from the changelog point of view, I suggest to group them separately (just like any other type).

> for `build`/`chore`/`ci`, i'm down reducing chore out and only allowing `build` and `ci` for specific differences, as i think anything that's a `chore` in this case would really be either a `build` or `ci` change, and `chore` could be eliminated entirely. I'm open to your opinion on this.

As you can see in the shared `commitlint.config.js`, I use `chore` with three scopes (`build`, `ci`, and `deps`, the last one used when updating the deps). I also use a scopeless `chore` type for other changes, e.g. when updating/adding `.gitignore` or `commitlint.config.js` or `update_dependencies.sh` and for initial commit (`chore: bootstrap Nx monorepo`). I also modified the MR (in GitLab; in GitLab, it is PR) message to the following:

```bash
chore: merge branch `$featBranch` to `dev`
```

Reply:
i'm open to the `feat`/`enh` distinction on release. For the sake of enabling that, i'll include the distinction in our config.

As for the `chore`/`build`/`ci` difference, i find angular's reasoning for removing chore and splitting it off to cover these cases...

one case of this... most CI already include skipping ci (how romcal does on `ci:` commits) via `[skip ci]`, which i think makes more sense.

`build`: covering build script changes, or just someone doing a manual build to deploy

`ci`: covering ci; such as actions, scripts for releases, configuration that doesn't directly tie to `build` updates

Reply:
changes/configs added for this, i'm considering adding `npm dedupe` as a "check" only if you modify the package-lock.json

Reply:
@tukusejssirs how do you prevent your commitlint from failing to load `selective-scope`? I have it as a dev dependency, but i'm not quite sure why it's being weird

Reply:
I have had issues with _loading_ the plugin. All I did is to install it and configure it in `commitlint.config.js`.

What it do when it fails to load? 🤔

Reply:
> As for the `chore`/`build`/`ci` difference, i find [angular's reasoning](https://gist.github.com/brianclements/841ea7bffdb01346392c?permalink_comment_id=3102984#gistcomment-3102984) for removing chore and splitting it off to cover these cases...

So, changes to `.gitignore` (and other, similar changes) are made under the `build` type, right? :thinking:

Reply:
> I have had issues with loading the plugin. All I did is to install it and configure it in commitlint.config.js.
> 
> What it do when it fails to load? 🤔

I've seen 
```diff
- Failed to load plugin commitlint-plugin-selective-scope.
```

so that's probably just a weirdness issue where i might just have to have `npm install` before lint-staged to make sure everything's enforced

Reply:
> I have **never** had issues with _loading_ the plugin. All I did is to install it and configure it in `commitlint.config.js`.

For some reason, I have missed the _never_ word in here. :thinking:

> so that's probably just a weirdness issue where i might just have to have `npm install` before lint-staged to make sure everything's enforced

It is still weird, because that error seems to come from `commitlint`, thus some packages need have been installed, while the plugin was still missing. :thinking:

Does it still happen to you? :thinking:

Reply:
interestingly, only after commits where a dedupe happens

Reply:
Maybe `npm dedupe` removes the plugin, as it can’t find it use anywhere in the code (`lib/`?). 🤔

Reply:
🤔

Reply:
hmm, trying to repro after manually deduping & running commitlint didn't cause issues for me

Reply:
i'm gonna chalk it up to me mistaking different machines that i was developing on at different times